File: 프로그래머스/4/60060. 가사 검색/가사 검색.java
import java.util.*;

class Solution {
    // 단어 길이별로 정방향과 역방향 단어를 저장하는 맵
    Map<Integer, List<String>> wordsByLength = new HashMap<>();
    Map<Integer, List<String>> reversedWordsByLength = new HashMap<>();

    public int[] solution(String[] words, String[] queries) {
        // 단어를 길이별로 분류하고 저장
        for (String word : words) {
            int length = word.length();
            wordsByLength.computeIfAbsent(length, k -> new ArrayList<>()).add(word);
            reversedWordsByLength.computeIfAbsent(length, k -> new ArrayList<>()).add(new StringBuilder(word).reverse().toString());
        }

        // 각 길이별 단어 리스트를 정렬
        for (List<String> wordList : wordsByLength.values()) {
            Collections.sort(wordList);
        }
        for (List<String> reversedWordList : reversedWordsByLength.values()) {
            Collections.sort(reversedWordList);
        }

        int[] answer = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            String query = queries[i];
            int length = query.length();

            // 해당 길이의 단어가 없으면 0 반환
            if (!wordsByLength.containsKey(length)) {
                answer[i] = 0;
                continue;
            }

            // 접두사가 '?'인 경우 (역방향 단어 리스트 사용)
            if (query.charAt(0) == '?') {
                String reversedQuery = new StringBuilder(query).reverse().toString();
                String left = reversedQuery.replace('?', 'a');
                String right = reversedQuery.replace('?', 'z');
                answer[i] = countByRange(reversedWordsByLength.get(length), left, right);
            } else { // 접미사가 '?'인 경우 (정방향 단어 리스트 사용)
                String left = query.replace('?', 'a');
                String right = query.replace('?', 'z');
                answer[i] = countByRange(wordsByLength.get(length), left, right);
            }
        }

        return answer;
    }

    // 이진 탐색을 사용해 [left, right] 범위에 속하는 단어의 개수를 반환
    private int countByRange(List<String> wordList, String left, String right) {
        int leftIndex = Collections.binarySearch(wordList, left);
        int rightIndex = Collections.binarySearch(wordList, right);

        // binarySearch는 요소를 찾지 못하면 음수 인덱스를 반환하므로, 삽입 위치를 계산
        if (leftIndex < 0) leftIndex = -(leftIndex + 1);
        if (rightIndex < 0) rightIndex = -(rightIndex + 1);
        else rightIndex += 1; // rightIndex가 정확히 일치하는 경우, 범위를 포함시키기 위해 +1

        return rightIndex - leftIndex;
    }
}

File: 프로그래머스/4/60060. 가사 검색/가사 검색.py
from collections import defaultdict
from bisect import bisect_left, bisect_right

def solution(words, queries):
    # 단어 길이별로 정방향과 역방향 단어 저장
    words_by_length = defaultdict(list)
    reversed_words_by_length = defaultdict(list)
    
    # 단어를 길이별로 분류하고 정렬
    for word in words:
        words_by_length[len(word)].append(word)
        reversed_words_by_length[len(word)].append(word[::-1])
    
    # 각 길이별 단어 리스트를 정렬
    for length in words_by_length:
        words_by_length[length].sort()
        reversed_words_by_length[length].sort()
    
    def count_by_range(words, left, right):
        """이진 탐색을 통해 [left, right] 범위에 속하는 단어의 개수를 반환"""
        return bisect_right(words, right) - bisect_left(words, left)
    
    answer = []
    for query in queries:
        length = len(query)
        
        # 해당 길이의 단어가 없으면 0을 반환
        if length not in words_by_length:
            answer.append(0)
            continue
        
        # 접두사가 '?'인 경우 -> 접미사를 확인 (뒤집은 단어 리스트 사용)
        if query[0] == '?':
            left = query[::-1].replace('?', 'a')
            right = query[::-1].replace('?', 'z')
            answer.append(count_by_range(reversed_words_by_length[length], left, right))
        else:  # 접미사가 '?'인 경우 -> 접두사를 확인 (정방향 단어 리스트 사용)
            left = query.replace('?', 'a')
            right = query.replace('?', 'z')
            answer.append(count_by_range(words_by_length[length], left, right))
    
    return answer